src/Common/CacheBase.h

```diff
@@ -1,55 +1,72 @@
 #pragma once
 
-#include <unordered_map>
-#include <list>
-#include <memory>
+#include <Common/Exception.h>
+#include <Common/ICachePolicy.h>
+#include <Common/LRUCachePolicy.h>
+#include <Common/SLRUCachePolicy.h>
+
+#include <atomic>
+#include <cassert>
 #include <chrono>
+#include <memory>
 #include <mutex>
-#include <atomic>
+#include <unordered_map>
 
 #include <Common/logger_useful.h>
 #include <base/defines.h>
 
 
 namespace DB
 {
-
-template <typename T>
-struct TrivialWeightFunction
+namespace ErrorCodes
 {
-    size_t operator()(const T &) const
-    {
-        return 1;
-    }
-};
-
+    extern const int BAD_ARGUMENTS;
+}
 
-/// Thread-safe cache that evicts entries which are not used for a long time.
+/// Thread-safe cache that evicts entries using special cache policy
+/// (default policy evicts entries which are not used for a long time).
 /// WeightFunction is a functor that takes Mapped as a parameter and returns "weight" (approximate size)
 /// of that value.
 /// Cache starts to evict entries when their total weight exceeds max_size.
 /// Value weight should not change after insertion.
 template <typename TKey, typename TMapped, typename HashFunction = std::hash<TKey>, typename WeightFunction = TrivialWeightFunction<TMapped>>
-class LRUCache
+class CacheBase
 {
 public:
     using Key = TKey;
     using Mapped = TMapped;
     using MappedPtr = std::shared_ptr<Mapped>;
 
-    /** Initialize LRUCache with max_size and max_elements_size.
-      * max_elements_size == 0 means no elements size restrictions.
-      */
-    explicit LRUCache(size_t max_size_, size_t max_elements_size_ = 0)
-        : max_size(std::max(static_cast<size_t>(1), max_size_))
-        , max_elements_size(max_elements_size_)
-        {}
+    /// TODO: Rewrite to custom struct with fields for all cache policies.
+    CacheBase(size_t max_size, size_t max_elements_size = 0, String cache_policy_name = "", double size_ratio = 0.5)
```

Comment:
There can't be a const variable now because if there will be no info about policy in config then we should change policy to default policy in the next 'if' statement. And if we will use reference without const then we can change reference value to default policy and I think it's not good.